Support external bootloader backends (RFC-0125)

grahamc opened the pull request:
###### Description of changes

Experimental implementation of https://github.com/NixOS/rfcs/pull/125.

To do:

- [x] Package the synthesis tool described in the RFC and as part of the documentation.

Not planned to be part of this PR:

- Migrating the existing bootloaders to use RFC-0125.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cole-h committed:
nixos/activation/bootspec: init bootspec support (RFC-0125)

cole-h committed:
nixos/boot/external: init

This allows supporting external bootloader backends.

Madouura commented:
I hope this isn't too out-of-scope, but I think it may be a good idea to add some fields for secure boot certificates/keys as well as a field for a hash of the current system closure for a TPM to verify against. Perhaps even some fields for kernel, bootloader, and any related hashes we could stick in a TPM and verify against, that aren't already included in the nixos system closure. (eg. kernel, bootloader, etc on /boot as ESP partition)
NixOS is very unique in that we can guarantee a certain set of software and data to be present and be what we are booting into, and we can update the TPM and secure boot keys/certs every ``nixos-rebuild``. The implementation of this goes out of scope, but I wanted to explain why adding those fields may be a good idea for a future case.

Madouura commented:
>Note that the RFC specifically mentions SecureBoot.

Yeah, that's why I was thinking it may be out-of-scope, I was thinking we might be able to use it to either forward some information to different bootloaders or use it for gathering information to implement such a scheme for verified boot.

>This is one of the very reasons it isn't appropriate to make those keys part of the system's closure: the policy of which keys and when they rotate is infinitely variable, and these parameters are very likely independent of a specific system closure

Okay, that's something I misunderstood. I didn't realize the bootspec document is part of the closure itself, in which case that throws my idea completely out, since we then can't guarantee things, thus ruining the entire point.
I'll have to think of a different method then, thank you for enlightening me.

K900 commented on nixos/modules/system/boot/loader/external/external.nix:
```diff
@@ -0,0 +1,38 @@
+    doc = ./external.xml;
+  };
+
+  options.boot.loader.external = {
```
I actually don't like the name here, considering in-tree bootloaders will probably want to use this as well. Maybe just have `boot.loader.installHook` everywhere and either ignore it if it's null or find some way to hook the old builders into it?

grahamc replied:
Why would they want to use it?

K900 replied:
Just for uniformity? Like, the way I see it, the final goal would be to just have all bootloader backends set this one option.

grahamc replied:
Do you mean they may want to use bootspec, or the `.external` namespace?

K900 replied:
I mean that they'd want to use bootspec and just set the `installer` option to their own installer instead of something user provided.

grahamc replied:
They'd be able to use bootspec and wouldn't need to use the "external" option. Every generation gets the bootspec file. This is defining the interface to how to specify an external bootloader. I think they can cohabitate via this option?

RaitoBezarius replied:
I propose we move this to another PR to track this thing once we have a more polished bootspec in-tree.

grahamc committed:
nixos/activation/bootspec: embed the entire contents of specialisation's bootspecs into the parent

See: https://github.com/NixOS/rfcs/pull/125#discussion_r871222614

grahamc committed:
nixos/activation/bootspec: embed the document into a bootspec subdir

cole-h commented on nixos/modules/system/activation/bootspec.nix:
```diff
@@ -0,0 +1,56 @@
+      generator =
+        let
+          specialisationLoader = (lib.mapAttrsToList
+            (childName: childToplevel: lib.escapeShellArgs [ "--slurpfile" childName "${childToplevel}/${filename}" ])
```
```suggestion
            (childName: childToplevel: lib.escapeShellArgs [ "--slurpfile" childName "${childToplevel}/bootspec/${filename}" ])
```

Also need to change the location here.

RaitoBezarius replied:
I think this was not solved? (hit that in SecureBoot)

SuperSandro2000 commented on nixos/modules/system/activation/bootspec.nix:
```diff
@@ -0,0 +1,56 @@
+            .init = $init
+            ' \
+            --sort-keys \
+            --arg toplevel "$out" \
+            --arg init "$out/init" \
```
I wouldn't mind a comment here what is being sorted. On first look we substitute top-level and init and then select specializations from the json.

RaitoBezarius replied:
@cole-h We still need to address this with a comment, I think it can be moved somewhere safe.

cole-h replied:
Open to suggestions :)

RaitoBezarius replied:
https://github.com/NixOS/nixpkgs/pull/172237/files#diff-097959078eaf4721c05782835844d922b4810ac16f8aec1598881036648ba36aR43

with the comment you dropped in the last commit, I think, would be a good place!

cole-h replied:
I have a better idea, coming soon (if it works). (What I don't  like about just moving it up there is that it's far from where the specialisation stuff is happening, so it's difficult to immediately know what it's talking about.)

cole-h replied:
[`bbada7c` (#172237)](https://github.com/NixOS/nixpkgs/pull/172237/commits/bbada7cb097e2986e2233b295f2dacae82b44350)

RaitoBezarius replied:
Great!

SuperSandro2000 commented on nixos/modules/system/boot/loader/external/external.nix:
```diff
@@ -0,0 +1,38 @@
+    boot.loader = {
+      grub.enable = mkDefault false;
+      systemd-boot.enable = mkDefault false;
+      supportsInitrdSecrets = false;
```
```suggestion
      supportsInitrdSecrets = mkDefault false;
```

SuperSandro2000 commented on nixos/doc/manual/from_md/configuration/bootloader-external.section.xml:
```diff
@@ -0,0 +1,41 @@
+<section xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" xml:id="sec-bootloader-external">
```
Is that file duplicated? I think so.

RaitoBezarius replied:
What makes you think so?

cole-h replied:
You're right, it's a dupe of https://github.com/NixOS/nixpkgs/pull/172237/files#diff-8639dc73210a20178e2d73b591a7c77402873cf0b32f58eed88d08493c7edea3, which should be automatically populated thanks to `meta.doc`. Fixed.

SuperSandro2000 commented on pkgs/tools/nix/bootspec/default.nix:
```diff
@@ -0,0 +1,24 @@
+}:
+rustPlatform.buildRustPackage {
+  pname = "bootspec";
+  version = "2022-05-31-unstable";
```
```suggestion
  version = "unstable-2022-05-31";
```

cole-h replied:
Don't we do `date-unstable` so that it can be version-sorted with everything else? Or am I misremembering?

K900 commented:
There's a small bug when using this on systems with no initrd, like nixos-on-wsl:

```
error: attribute 'initialRamdisk' missing

       at /nix/store/c9c7d2zfmjwcx3pmiygh06ggyhhwynwf-nixpkgs-patched/nixos/modules/system/activation/bootspec.nix:19:27:

           18|               kernelParams = config.boot.kernelParams;
           19|               initrd = "${config.system.build.initialRamdisk}/${config.system.boot.loader.initrdFile}";
             |                           ^
           20|               initrdSecrets = "${config.system.build.initialRamdiskSecretAppender}/bin/append-initrd-secrets";

       … while evaluating the attribute 'text' of the derivation 'boot.v1.json'

       at /nix/store/c9c7d2zfmjwcx3pmiygh06ggyhhwynwf-nixpkgs-patched/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

       … while evaluating the attribute 'v1.generator'

       at /nix/store/c9c7d2zfmjwcx3pmiygh06ggyhhwynwf-nixpkgs-patched/nixos/modules/system/activation/bootspec.nix:24:7:

           23|
           24|       generator =
             |       ^
           25|         let

       … while evaluating the attribute 'writer'

       at /nix/store/c9c7d2zfmjwcx3pmiygh06ggyhhwynwf-nixpkgs-patched/nixos/modules/system/activation/bootspec.nix:55:3:

           54|   # the path of the `toplevel` itself and the `init` executable).
           55|   writer = schemas.v1.generator;
             |   ^
           56| }

       … while evaluating the attribute 'buildCommand' of the derivation 'nixos-system-sayaka-wsl-22.11pre-git'

       at /nix/store/c9c7d2zfmjwcx3pmiygh06ggyhhwynwf-nixpkgs-patched/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let
```

K900 committed:
nixos/activation: don't generate bootspec for containers

RaitoBezarius commented on nixos/doc/manual/from_md/configuration/bootloader-external.section.xml:
```diff
@@ -0,0 +1,41 @@
+    tooling using the <quote>external</quote> bootloader option.
+  </para>
+  <para>
+    Imagine you have created a new packaged called FooBoot. FooBoot
```
package

K900 commented:
Are the kernel commits supposed to be here? They conflict with master: https://github.com/NixOS/nixpkgs/commit/23bda1f2db161501a68e119c402df166c35c9545

RaitoBezarius committed:
nixos/activation/bootspec: fix slurping specialisation bootspecs

RaitoBezarius committed:
nixos/activation/bootspec: add @raitobezarius as a code-owner

RaitoBezarius committed:
nixos/activation/bootspec: module-ify

This does the following:

* turns bootspec into a NixOS module
* validates bootspecs with Cue
* exposes internal knobs

RaitoBezarius committed:
nixos/tests/bootspec: init

This will test various scenarios of bootspec generation.

RaitoBezarius committed:
nixos/boot/external: add @raitobezarius as maintainer and allow initrd secrets

RaitoBezarius committed:
nixos/activation/bootspec: remove SB extension example in Cue schema

RaitoBezarius committed:
nixos/activation/bootspec: add bootspec chapter in NixOS manual

RaitoBezarius committed:
nixos/activation/bootspec: add some comments to explain the delicate manipulations

RaitoBezarius commented:
@grahamc @cole-h Do we still need this synthesis package tool for the PR? Now there is a experimental flag.
I think we can start the review process, cc @lheckemann

cole-h committed:
nixos/boot/external: DocBook -> Markdown

RaitoBezarius commented:
> There's a small bug when using this on systems with no initrd, like nixos-on-wsl:

Does it even make sense to have a bootspec on them?

cole-h committed:
nixos/boot/external: fixup typo in generated docs, regenerate docs

cole-h committed:
nixos/activation/bootspec: DocBook -> Markdown, add description for extensions field

linyinfeng commented on nixos/modules/system/activation/bootspec.nix:
```diff
@@ -0,0 +1,111 @@
+            --arg init "$out/init" \
+            < ${json} \
+            | ${pkgs.jq}/bin/jq \
+              --sort-keys \ # Slurp all specialisations and inject them as values in .specialisations.{name} = {specialisation bootspec}.
```
The comment causes the following build error:

```
jq: error: Could not open file Slurp: No such file or directory
jq: error: Could not open file all: No such file or directory
jq: error: Could not open file specialisations: No such file or directory
jq: error: Could not open file and: No such file or directory
jq: error: Could not open file inject: No such file or directory
jq: error: Could not open file them: No such file or directory
jq: error: Could not open file as: No such file or directory
jq: error: Could not open file values: No such file or directory
jq: error: Could not open file in: No such file or directory
jq: error: Could not open file .specialisations.{name}: No such file or directory
jq: error: Could not open file =: No such file or directory
jq: error: Could not open file {specialisation: No such file or directory
jq: error: Could not open file bootspec}.: No such file or directory
```

RaitoBezarius replied:
Ugh, thank you !

RaitoBezarius commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -79,6 +79,11 @@ let
 
       echo -n "${toString config.system.extraDependencies}" > $out/extra-dependencies
 
+      ${optionalString (!config.boot.isContainer) ''
```
Ha, I forgot to defend this with `config.boot.bootspec.enable` too.

cole-h committed:
nixos/activation/bootspec: drop problematic comment, only generate bootspec when bootspec is enabled

cole-h committed:
nixos/activation/bootspec: refactor the generator script

We separate the different steps (injecting the toplevel and injecting
the specialisations) so that it's easy to document what each snippet is
actually doing.

RaitoBezarius commented:
I think it's in a good shape now, @cole-h @grahamc would you agree to make it "ready for review" ?

cole-h committed:
nixos/boot/external: drop duplicated external bootloader documentation

The documentation is generated thanks to `meta.doc`, and was out of
date anyways.

cole-h committed:
bootspec: init at unstable-2022-12-05

The bootspec package provides synthesis and validation tooling for
RFC-0125 compliant bootspec documents.

cole-h committed:
nixos/activation/bootspec: fixup improper $out substitution

cole-h commented:
I cleaned up the history a little bit (mostly renaming commits to more accurately describe what area they're touching, i.e. `bootspec` -> `nixos/activation/bootspec` in most cases), completed the TODO in the PR body of packaging the bootspec synthesize tool, and fixed the fact that `$out` being `escapeShellArgs`'d will, of course, not let it expand to the `$out` we expect (switched to `placeholder "out"`, which Nix will handle for us; no need to hack around it).

I am pleased to mark this PR as ready for review :)

cole-h committed:
nixos/activation/bootspec: bootspec owners also own the cue file

Mic92 commented on nixos/doc/manual/from_md/development/bootspec.chapter.xml:
```diff
@@ -0,0 +1,73 @@
+  <title>Experimental feature: Bootspec</title>
+  <para>
+    Bootspec is a experimental feature, introduced in the
+    <link xlink:href="https://github.com/NixOS/rfcs/125">RFC-0125
```
```suggestion
    <link xlink:href="https://github.com/NixOS/rfcs/pull/125">RFC-0125
```

RaitoBezarius commented:
@cole-h Can you rebase the branch please? Thanks! (so we can remove the Cue stuff now it's merged in nixpkgs)

Mic92 commented on nixos/tests/bootspec.nix:
```diff
@@ -0,0 +1,143 @@
+
+    testScript = ''
+      machine.start()
+      machine.wait_for_unit("multi-user.target")
```
This test hangs on my system.

RaitoBezarius replied:
All that's missing is: `boot.loader.grub.efiSupport = true;` in `grub` and it should work.

RaitoBezarius replied:
fixed in https://github.com/DeterminateSystems/nixpkgs/pull/5 @cole-h

RaitoBezarius committed:
nixos/tests/bootspec: add EFI support for GRUB test

RaitoBezarius commented:
As this is a PR guarded by an experimental flag and multiple tests and one which is in good shape, I think I can self-merge it in the next days once the GRUB fix is landed. Plus, it is easy to revert and enable developers to test SecureBoot (and more!) properly.

SuperSandro2000 commented on nixos/modules/system/activation/bootspec.nix:
```diff
@@ -0,0 +1,124 @@
+  };
+
+  config = lib.mkIf (cfg.enable) {
+    warnings = [
+      ''RFC-0125 is not merged yet, this is a feature preview of bootspec.
+        The schema is not definitive and features are not guaranteed to be stable until RFC-0125 is merged.
+        See:
+        - https://github.com/NixOS/nixpkgs/pull/172237 to track merge status in nixpkgs.
+        - https://github.com/NixOS/rfcs/pull/125 to track RFC status.
+      ''
+    ];
```
There is no native way to turn this rather annoying warning off other than introducing another option?

RaitoBezarius replied:
An other way to turn it off is to merge RFC-0125 :)
